python/packages/ai/teams/app.py

```diff
@@ -298,28 +317,26 @@ async def _on_turn(self, context: TurnContext):
 
         try:
             await typing.start(context)
+            state: TurnState
 
-            state = await self._options.turn_state_manager.load_state(
-                self._options.storage, context
-            )
+            if self._turn_state_factory:
+                state = await self._turn_state_factory(context.activity)
+            else:
+                state = await self._load_state(context.activity)
 
             # run before turn middleware
             for before_turn in self._before_turn:
-                if not await before_turn(context, state):
-                    await self._options.turn_state_manager.save_state(
-                        self._options.storage, context, state
-                    )
+                is_ok = await before_turn(context, cast(StateT, state))
+                await state.save(self._options.storage)
```

Comment:
Seems this will save the state after every before_turn handler's execution, instead of only save when before_turn handler failed. Do we need some optimization here?

Reply:
I did it this way to ensure that the state is always persisted between route handlers, but now that the `TurnState` has its own `save` method they can technically save it whenever they want. I'll ask around and see what everyone else thinks and we can discuss in our next meeting as well.